modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -421,66 +421,141 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
                         }
                     }
                 }
-                int prevNodeLength = 0;
-                boolean isPrevNodeExcluded = false;
-                for (TextRun r: runs) {
-                    if (!r.getTextSpan().getText().equals(text) || (r.getStart() < textRunStart && r.getTextSpan().getText().equals(text))) {
-                        prevNodeLength += r.getWidth();
-                        continue;
+
+                BaseBounds textBounds = new BoxBounds();
+                if (isMirrored) {
+                    int runIdx = 0;
+                    for (TextRun r: runs) {
+                        if (r.getStart() == curRunStart) {
+                            run = r;
+                            break;
+                        }
+                        runIdx++;
                     }
-                    if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)) {
-                        BaseBounds textBounds = new BoxBounds();
-                        getBounds(r.getTextSpan(), textBounds);
-                        if (textBounds.getMinX() == 0 && !isPrevNodeExcluded) {
-                            x -= prevNodeLength;
-                            isPrevNodeExcluded = true;
+
+                    boolean textFound = false;
+                    for (int i = 0; i <= runIdx; i++) {
+                        TextRun r = runs[i];
+                        if (r.getStart() != curRunStart && r.getTextSpan().getText().equals(text)
+                                && x > r.getWidth() && textWidthPrevLine == 0) {
+                            x -= r.getWidth();
+                            textFound = true;
+                            continue;
+                        }
+                        if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)
+                                && r.getStart() == curRunStart) {
+                            if (x > r.getWidth() || textWidthPrevLine > 0) {
+                                getBounds(r.getTextSpan(), textBounds);
+                                x -= (run.getLocation().x - textBounds.getMinX());
+                            }
+                            break;
                         }
-                        if (x > r.getWidth()) {
+                        /* This condition handles LTR Text nodes present between
+                           a Text node containing both LTR and RTL text. */
+                        if (!r.getTextSpan().getText().equals(text) && textFound
+                                && x > r.getWidth() && r.getStart() < curRunStart) {
                             x -= r.getWidth();
-                            relIndex += r.getLength();
+                        }
+                    }
+                    for (int i = runs.length - 1; i > runIdx; i--) {
+                        TextRun r = runs[i];
+                        boolean addLtrIdx = run.getTextSpan().getText().length() != run.length;
```

Comment:
Q: we should never get an NPE here, correct?

Reply:
Yes we will not get NPE here. We enter this block only when `spans != null` so I don't think we need another null check here.
We have one null check at l.no.515 which is not really necessary but I will keep it as it is for now.